### ISSUE TITLE
Scanning from HEAD to first commit of feature branch scans the entire history up to HEAD.

### ISSUE THREAD

**Issue by eternicode** [zricethezav/gitleaks#407]
**Describe the bug**

In our team, when we develop a new feature, we create the feature branch off of master, and then have individual sub-feature branches off of the feature branch where individuals do their work. This normally results in a feature branch that is nothing but merge commits, which is then merged into master.

We are trying to set up gitleaks to scan each branch from branch head to where it branches off from master (ideally, this would be the first commit to the first sub-branch that was merged into the feature branch). For our feature branches, however, this results in gitleaks scanning almost the entire repository history before finding the `--commit-to` hash.

**To Reproduce**

The following bash script, when called with a directory location, creates a repository at said location, with a history that can replicate this issue.

`make_leaker.sh leaker/`

```bash
#!/bin/bash
# make_leaker.sh

REPO="$1"

if [ -z "$REPO" ]; then
    echo "First argument must be new repo location" >&2
    exit 1
fi

git init "$REPO"
cd "$REPO"

COMMIT=1

make_commit() {
    msg="$1"

    branch_name="$(git symbolic-ref HEAD 2>/dev/null)" || branch_name="unnamed_branch"
    branch_name=${branch_name##refs/heads/}
    file="${branch_name}_file.txt"

    echo "$COMMIT" > $file
    ((COMMIT+=1))
    git add $file
    git commit -m "$msg"
    # Put at least 1 second between each commit to keep the git graphlog pretty
    sleep 1
}

make_branch() {
    name="$1"
    git checkout -b "$name"
}

get_branch() {
    name="$1"
    git checkout "$name"
}

make_commit "master 1"
make_commit "master 2"
make_commit "master 3"
make_commit "master 4"
make_commit "master 5"
make_commit "master 6"

make_branch feature_branch
make_branch sub_feature_branch
make_commit "branch2 1"
make_commit "branch2 2"
make_commit "branch2 3"

get_branch feature_branch
git merge --no-ff -m "merge sub_feature_branch" sub_feature_branch

get_branch master
make_commit "master 7"
make_commit "master 8"
make_commit "master 9"
```

The repository will have a history that looks like this:

```
$ git log --no-color --graph --pretty=format:'changeset:   %H%d %an <%ae> %ci%nsummary:     %s%n' --all
* changeset:   4a80e4e15c72d58bd81ffc8f13cf0fcadf8a70dc (HEAD -> master) Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:26 -0400
| summary:     master 9
| 
* changeset:   b98ae921f109c6837b4cc95fc79560dd549f098f Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:25 -0400
| summary:     master 8
| 
* changeset:   6ea2ed9ac4a1973234392e2c5bc04245b4ea0fe1 Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:24 -0400
| summary:     master 7
|     
| *   changeset:   a9f16cb8d8d5de499c877df95782f70237ee91f1 (feature_branch) Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:24 -0400
| |\  summary:     merge sub_feature_branch
|/ /  
| | 
| * changeset:   d853e6d7453630aad8a3d5df4377db8f84a86e26 (sub_feature_branch) Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:23 -0400
| | summary:     branch2 3
| | 
| * changeset:   2e5c124fac798e07df1128e0d5abf88aace7d95a Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:22 -0400
| | summary:     branch2 2
| | 
| * changeset:   786dff1eccc6c0cd9b807a3d019bb2bd985a34ee Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:21 -0400
|/  summary:     branch2 1
| 
* changeset:   6332a4302192c8651b46bac03e518d30e492b27f Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:20 -0400
| summary:     master 6
| 
* changeset:   00fae103eca74cb75d962edee9d958ab005635b7 Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:19 -0400
| summary:     master 5
| 
* changeset:   58d4d8d766b332a3d539f3de01ed81cc12980c9d Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:18 -0400
| summary:     master 4
| 
* changeset:   1a3c08855e42ca46aee6a8160e1b39dc9318b068 Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:17 -0400
| summary:     master 3
| 
* changeset:   df00fcbdd5d99a507c2b613fab556abbd4737917 Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:16 -0400
| summary:     master 2
| 
* changeset:   71e63ac13e3ee23264bb7e977f0a973689bd16c5 Andrew Rowls <andrew@eternicode.com> 2020-07-09 10:08:15 -0400
  summary:     master 1
```

To scan `feature_branch`, we use `git rev-list master..feature_branch | tail -1`, which yields the hash `786dff1eccc6c0cd9b807a3d019bb2bd985a34ee` (the first commit on `sub_feature_branch`), and subsequently pass that hash to gitleaks as the `--commit-to`, as well as `--branch=feature_branch`.

```
docker run --rm --name=gitleaks --volume "$(pwd)":/code zricethezav/gitleaks --verbose --pretty --repo-path=/code/ --commit-to="$(git rev-list master..feature_branch | tail -1)" --branch=feature_branch
```

While we would like this to result in four commits scanned, we get this line at the end:

```
INFO[2020-07-09T14:18:01Z] No leaks detected. 10 commits audited in 5 milliseconds 364 microseconds
```

Indicating that gitleaks has scanned the above four commits, as well as the six commits before that on `master`.

*Furthermore...*

Using the most recent common parent, `6332a430`, causes gitleaks to stop before it has scanned any commits in the feature branches:

```
$ docker run --rm --name=gitleaks --volume "$(pwd)":/code zricethezav/gitleaks --verbose --pretty --repo-path=/code/ --commit-to="6332a4302192c8651b46bac03e518d30e492b27f" --branch=feature_branch
INFO[2020-07-09T15:51:19Z] No leaks detected. 2 commits audited in 1 millisecond 338 microseconds 
```

**Expected behavior**

In the first case, ideally, gitleaks will scan the commits `a9f16cb8`, `d853e6d7`, `2e5c124f`, and `786dff1e`, and stop.

Minimally, I'd expect gitleaks to see `786dff1e` and stop long before it reaches `71e63ac1`, the beginning of the history.

In the second case, I'd expect gitleaks to scan all 5 commits between the `feature_branch` HEAD and common parent `6332a430`, inclusive.

**Basic Info (please complete the following information):**
 - OS: ubuntu linux
 - Gitleaks Version: v4.3.1

cc @zricethezav

**Comment by zricethezav** [comment]
Gitleaks uses go-git as it's git library and doesn't provide a branch filter. Instead it will [attempt to find the reference with the branch name and use that commit as the starting point](https://github.com/zricethezav/gitleaks/blob/master/audit/util.go#L444-L470). The problem with this is branches that have been merged in create a merge commit which has two parents. Git log will traverse the commit prior to branching first, then the commits within the branch before continuing. This is why you're getting more commits scanned than you intend. 

The inclusion of `a9f16cb8d8d5de499c877df95782f70237ee91f1` in your case is the problem. Adding the `branch` option is somewhat contentious and there is an argument that it should be removed as it will always be overzealous when scanning.

I recommend you manually set the `commit-to` and `commit-from` as is demonstrated in [gitleaks-action](https://github.com/zricethezav/gitleaks-action/blob/master/entrypoint.sh#L20-L22).

I'll be looking more into this but that's what I got for now.

cc @eternicode
